File: conway-ruby/lib/conway/game_board.rb
module Conway
  class GameBoard
    attr_reader :grid

    def initialize(max_grid_size)
      @max_grid_size = max_grid_size
      @grid = Array.new(@max_grid_size) { Array.new(@max_grid_size, ' ') }
    end

    def initialize_pattern(input_pattern, start_row, start_col)
      (0..input_pattern.length - 1).each do |row|
        (0..input_pattern[row].length - 1).each do |col|
          @grid[row + start_row][col + start_col] = input_pattern[row][col]
        end
      end
    end

    def number_of_live_adjacent_cells(x_pos, y_pos)
      number_of_live_cells = 0
      number_of_live_cells += number_live_cells_top_row(x_pos, y_pos)
      number_of_live_cells += number_live_cells_same_row(x_pos, y_pos)
      number_of_live_cells += number_live_cells_bottom_row(x_pos, y_pos)
      number_of_live_cells
    end

    def update_grid(temporary_grid)
      (0..@max_grid_size - 1).each do |row|
        (0..@max_grid_size - 1).each do |col|
          @grid[row][col] = temporary_grid[row][col]
        end
      end
    end

    def number_live_cells_top_row(x_pos, y_pos)
      number_of_live_neighbors = 0
      number_of_live_neighbors += 1 if x_pos - 1 >= 0 && y_pos - 1 >= 0 && cell_alive?(x_pos - 1, y_pos - 1)
      number_of_live_neighbors += 1 if x_pos - 1 >= 0 && cell_alive?(x_pos - 1, y_pos)
      number_of_live_neighbors += 1 if x_pos - 1 >= 0 && y_pos + 1 < @max_grid_size && cell_alive?(x_pos - 1, y_pos + 1)
      number_of_live_neighbors
    end

    def number_live_cells_bottom_row(x_pos, y_pos)
      number_of_live_neighbors = 0
      number_of_live_neighbors += 1 if x_pos + 1 < @max_grid_size && y_pos - 1 >= 0 && cell_alive?(x_pos + 1, y_pos - 1)
      number_of_live_neighbors += 1 if x_pos + 1 < @max_grid_size && cell_alive?(x_pos + 1, y_pos)
      if x_pos + 1 < @max_grid_size && y_pos + 1 < @max_grid_size && cell_alive?(x_pos + 1, y_pos + 1)
        number_of_live_neighbors += 1
      end
      number_of_live_neighbors
    end

    def number_live_cells_same_row(x_pos, y_pos)
      number_of_live_neighbors = 0
      number_of_live_neighbors += 1 if y_pos - 1 >= 0 && cell_alive?(x_pos, y_pos - 1)
      number_of_live_neighbors += 1 if y_pos + 1 < @max_grid_size && cell_alive?(x_pos, y_pos + 1)
      number_of_live_neighbors
    end

    def cell_alive?(x_pos, y_pos)
      @grid[x_pos][y_pos] == '#'
    end

    private :number_live_cells_bottom_row, :number_live_cells_top_row, :number_live_cells_same_row, :cell_alive?
  end
end

File: conway-java/src/main/java/com/gojek/pasti/conway/GameBoard.java
package com.gojek.pasti.conway;

import java.util.ArrayList;

public class GameBoard {
  private char[][] grid;
  private final int maxGridSize;

  public GameBoard(int maxGridSize) {
    grid = new char[maxGridSize][maxGridSize];
    this.maxGridSize = maxGridSize;
    createEmptyGridWithDeadCells();
  }

  protected char[][] getCurrentGrid() {
    return grid;
  }

  private void createEmptyGridWithDeadCells() {
    for (int row = 0; row < maxGridSize; row++) {
      for (int col = 0; col < maxGridSize; col++) {
        grid[row][col] = ' ';
      }
    }
  }

  protected void initializePattern(
      ArrayList<ArrayList<Character>> inputPattern, int startX, int startY) {
    for (int row = 0; row < inputPattern.size(); row++) {
      for (int col = 0; col < inputPattern.get(row).size(); col++) {
        grid[startX + row][startY + col] = inputPattern.get(row).get(col);
      }
    }
  }

  protected int numberOfLiveAdjacentCells(int x, int y) {
    int numberOfLiveNeighbors = 0;
    if (x - 1 >= 0 && y - 1 >= 0 && isCellAlive(x - 1, y - 1)) numberOfLiveNeighbors += 1;
    if (x - 1 >= 0 && isCellAlive(x - 1, y)) numberOfLiveNeighbors += 1;
    if (x - 1 >= 0 && y + 1 < maxGridSize && isCellAlive(x - 1, y + 1)) numberOfLiveNeighbors += 1;
    if (y - 1 >= 0 && isCellAlive(x, y - 1)) numberOfLiveNeighbors += 1;
    if (y + 1 < maxGridSize && isCellAlive(x, y + 1)) numberOfLiveNeighbors += 1;
    if (x + 1 < maxGridSize && y - 1 >= 0 && isCellAlive(x + 1, y - 1)) numberOfLiveNeighbors += 1;
    if (x + 1 < maxGridSize && isCellAlive(x + 1, y)) numberOfLiveNeighbors += 1;
    if (x + 1 < maxGridSize && y + 1 < maxGridSize && isCellAlive(x + 1, y + 1))
      numberOfLiveNeighbors += 1;
    return numberOfLiveNeighbors;
  }

  private boolean isCellAlive(int x, int y) {
    return grid[x][y] == '#';
  }

  protected void updateGrid(char[][] temporaryBoard) {
    for (int row = 0; row < maxGridSize; row++) {
      System.arraycopy(temporaryBoard[row], 0, grid[row], 0, maxGridSize);
    }
  }
}
